db/types.go

```diff
@@ -0,0 +1,198 @@
+	Get([]byte) ([]byte, error)
+
+	// Has checks if a key exists.
+	// CONTRACT: key, value readonly []byte
+	Has(key []byte) (bool, error)
+
+	// Iterator returns an iterator over a domain of keys, in ascending order. The caller must call
+	// Close when done. End is exclusive, and start must be less than end. A nil start iterates
+	// from the first key, and a nil end iterates to the last key (inclusive). Empty keys are not
+	// valid.
+	// CONTRACT: No writes may happen within a domain while an iterator exists over it.
+	// CONTRACT: start, end readonly []byte
+	Iterator(start, end []byte) (Iterator, error)
+
+	// ReverseIterator returns an iterator over a domain of keys, in descending order. The caller
+	// must call Close when done. End is exclusive, and start must be less than end. A nil end
+	// iterates from the last key (inclusive), and a nil start iterates to the first key (inclusive).
+	// Empty keys are not valid.
+	// CONTRACT: No writes may happen within a domain while an iterator exists over it.
+	// CONTRACT: start, end readonly []byte
+	ReverseIterator(start, end []byte) (Iterator, error)
+
+	// Discards the transaction, invalidating any future operations on it.
+	Discard()
+}
+
+// DBWriter is a write-only transaction interface.
+// It is safe for concurrent writes, following an optimistic (OCC) strategy, detecting any write
+// conflicts and returning an error on commit, rather than locking the DB.
+// This can be used to wrap a write-optimized batch object if provided by the backend implementation.
+type DBWriter interface {
+	// Set sets the value for the given key, replacing it if it already exists.
+	// CONTRACT: key, value readonly []byte
+	Set([]byte, []byte) error
+
+	// Delete deletes the key, or does nothing if the key does not exist.
+	// CONTRACT: key readonly []byte
+	Delete([]byte) error
+
+	// Flushes pending writes and discards the transaction.
+	Commit() error
+
+	// Discards the transaction, invalidating any future operations on it.
+	Discard()
+}
+
+// DBReadWriter is a transaction interface that allows both reading and writing.
+type DBReadWriter interface {
+	DBReader
+	DBWriter
+}
+
+// Iterator represents an iterator over a domain of keys. Callers must call Close when done.
+// No writes can happen to a domain while there exists an iterator over it, some backends may take
+// out database locks to ensure this will not happen.
+//
+// Callers must make sure the iterator is valid before calling any methods on it, otherwise
+// these methods will panic. This is in part caused by most backend databases using this convention.
+//
+// As with DBReader, keys and values should be considered read-only, and must be copied before they are
+// modified.
+//
+// Typical usage:
+//
+// var itr Iterator = ...
+// defer itr.Close()
+//
+// for ; itr.Valid(); itr.Next() {
+//   k, v := itr.Key(); itr.Value()
+//   ...
+// }
+// if err := itr.Error(); err != nil {
+//   ...
+// }
+type Iterator interface {
```

Comment:
I've written/used iterators with the pattern:

```go

var iter Iterator

for iter.Next() {

}

if err := iter.Err(); err != nil {
    return err
}
if err := iter.Close(); err != nil {
    return err 
}
```

Which collapses Valid/Next,

Reply:
Seems fine, the only odd thing about this is it requires the iterator to start in an invalid state

Reply:
:+1:  - had similar opinion about it above.

Reply:
Thinking a bit more on this, it may not be ideal as it violates the single-responsibility principle. The interface is smaller but you can't check the iterator's state without mutating it.

That said, I have pushed the change to the interface.